### ISSUE TITLE
Unable to pull image using 'Pull Once - Update always' scheme

### ISSUE THREAD

**Issue by djl197** [jenkinsci/docker-plugin#339]
We have just upgraded to plugin version: 0.15.0
When we try to provision a node we get the following:

<pre>Oct 07, 2015 4:42:09 PM SEVERE com.nirima.jenkins.plugins.docker.DockerCloud$1 call
Error in provisioning; template='DockerTemplate{configVersion=2, labelString='docker cap_builder', launcher=com.nirima.jenkins.plugins.docker.launcher.DockerComputerSSHLauncher@21460684, remoteFsMapping='', remoteFs='/home/jenkins', instanceCap=8, mode=EXCLUSIVE, retentionStrategy=com.nirima.jenkins.plugins.docker.strategy.DockerOnceRetentionStrategy@671e7b79, numExecutors=1, dockerTemplateBase=DockerTemplateBase{image=docker-registry.domain.com:5000/ubuntu_14.04_cap_builder:latest}, removeVolumes=false, pullStrategy=PULL_LATEST}' for cloud='dev-bamboo'
com.github.dockerjava.api.DockerClientException: Could not pull image: null
    at com.github.dockerjava.core.command.PullImageResultCallback.awaitSuccess(PullImageResultCallback.java:46)
    at com.nirima.jenkins.plugins.docker.DockerCloud.pullImage(DockerCloud.java:341)
    at com.nirima.jenkins.plugins.docker.DockerCloud.provisionWithWait(DockerCloud.java:348)
    at com.nirima.jenkins.plugins.docker.DockerCloud.access$000(DockerCloud.java:58)
    at com.nirima.jenkins.plugins.docker.DockerCloud$1.call(DockerCloud.java:238)
    at com.nirima.jenkins.plugins.docker.DockerCloud$1.call(DockerCloud.java:235)
    at jenkins.util.ContextResettingExecutorService$2.call(ContextResettingExecutorService.java:46)
    at java.util.concurrent.FutureTask.run(FutureTask.java:262)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(Thread.java:745)</pre>

It looks like its attempting to pull but failing.  Do you know why this might be happening.
The other slight difference is this is pointed to a swarm (docker-swarm) - might that be confusing the issue? I.E not yet supported?

Dan

**Comment by djl197** [comment]
Looks like its related to: 
https://github.com/docker-java/docker-java/pull/337

So If that gets fixed a bump of docker-java might be good

**Comment by KostyaSha** [comment]
Hi, i planned fix this on this weekend.

**Comment by hawky-4s-** [comment]
We did hit the same bug, that's why I created https://github.com/docker-java/docker-java/pull/337 :shipit:

**Comment by KostyaSha** [comment]
@hawky-4s- oh, i thought i get the same issue without swarm. Trying reproduce. But i will back tomorrow to code.

**Comment by schneidexe** [comment]
I also get this issue without swarm on docker 1.8.3

**Comment by magnayn** [comment]
Do you have a stacktrace (pref with logging set to DEBUG) ?

The swarm UC fails because of docker-java not recognising the response. docker-normal should work..

**Comment by schneidexe** [comment]
I get the same stacktrace like in the opening post. DEBUG does not provide more info...

**Comment by magnayn** [comment]
​Then probably 1.8.3 has, yet again, changed the responses.

docker-java looks for responses containing "Download complete", "Image is
up to date" and "Downloaded newer image" to determine success.

If you do a pull manually, you may be able to figure what the correct value
is...
​

**Comment by schneidexe** [comment]
With docke r1.8.1 it's still working. Checking 1.8.2 now ... changelog (https://github.com/docker/docker/blob/master/CHANGELOG.md) says a few things about docker pull in 1.8.2. So I guess it will also not work...

**Comment by schneidexe** [comment]
1.8.3 spit out a warning "<blah/blub>: this image was pulled from a legacy registry.  Important: This registry version will not be supported in future versions of docker." on the last line. Might that be the cause? But the line before it says "Status: Downloaded newer image for <blah/blub>"...

**Comment by magnayn** [comment]
Possibly. Might be the "Status: " bit that's new.

**Comment by schneidexe** [comment]
But the "Status: " thing is also there in 1.8.1 and with this version it works...

**Comment by magnayn** [comment]
Ah - it the "Status: Downloaded newer image" the *last* line ?

PullImageResultCallback from docker-java only considers the last item.

If you're feeling brave, you could hand-patch your plugin. something like (untested):
```diff
@@ -338,7 +340,26 @@ public class DockerCloud extends Cloud {

             long startTime = System.currentTimeMillis();
             //Identifier amiId = Identifier.fromCompoundString(ami);
-            getClient().pullImageCmd(imageName).exec(new PullImageResultCallback()).awaitSuccess();
+            getClient().pullImageCmd(imageName).exec(new PullImageResultCallback(){
+                boolean success = false;
+
+                public void onNext(PullResponseItem item) {
+                    success = success | item.isPullSuccessIndicated();
+                    LOGGER.debug(item.toString());
+                }
+
+                public void awaitSuccess() {
+                    try {
+                        this.awaitCompletion();
+                    } catch (InterruptedException var2) {
+                        throw new DockerClientException("", var2);
+                    }
+
+                    if(!success) {
+                        throw new DockerClientException("Could not pull image");
+                    }
+                }
+            }).awaitSuccess();
             long pullTime = System.currentTimeMillis() - startTime;
             LOGGER.info("Finished pulling image '{}', took {} ms", imageName, pullTime);
         }
```

**Comment by schneidexe** [comment]
Thanks for pointing that out! In case you use a v1 registry the last line is the warning which seems to cause that the result get not parsed properly. I will see if I can patch that thing. Other option I'll check is using a v2 registry, since v1 will anyway not be supported by future docker versions.

**Comment by magnayn** [comment]
Putting my brain into gear, actually even simpler would be to just replace the
```
.awaitSuccess()
```

call with 

```
.awaitCompletion()
```

Since a) this 'detect what docker said to determine success' seems to be v. flaky and breaks on (minor) verion number bumps and b) actually, if it failed you'll find out soon enough as the image won't be there.

**Comment by schneidexe** [comment]
:+1: I like this solution. I was actually able to fix the issue now by upgrading our repo to v2. Anyway I think that should be fixed in the plugin to be more resilient.

**Comment by fraz3alpha** [comment]
Did the fix mentioned above (https://github.com/jenkinsci/docker-plugin/issues/339#issuecomment-152112868) ever get implemented?

**Comment by magnayn** [comment]
There is an option to ignore the return response in the latest rel.

**Comment by ndeloof** [comment]
fixed by 847f63730e41387b2abdf90e0674db9593b660d5
we check image pull was successful as the image is well there, not by trying to parse progress messages.
